### PR TITLE
MM-16888 - Fix canonical vs. migration mismatches

### PR DESCRIPTION
**Pull request by reflog**
#### Summary

Make sure test-db-migrations works and added it to Jenkins to enforce it in the future

#### Ticket Link
Fixes: https://mattermost.atlassian.net/browse/MM-16888

### CODE REVIEW

**Comment by reflog** [comment]
I've noticed that adding make test-db-migrations to Jenkins.pr didn't work:
```
exit status 102

Makefile:334: recipe for target 'run-cli' failed

make[1]: *** [run-cli] Error 1

make[1]: Leaving directory '/go/src/github.com/mattermost/mattermost-server'

Generating dump

./scripts/psql-migration-test.sh: 26: ./scripts/psql-migration-test.sh: docker: not found

./scripts/psql-migration-test.sh: 27: ./scripts/psql-migration-test.sh: docker: not found

Removing databases created for db comparison

./scripts/psql-migration-test.sh: 30: ./scripts/psql-migration-test.sh: docker: not found

Generating diff

Both schemas are same
```


@cpanato - any suggestions on how to fix this?

**Comment by mkraft** [comment]
@reflog Can you try `make nuke && make test-db-migration`? Your PR runs for me successfully so I wonder if you got into a state where the docker container names were wrong because of some recent changes in that area.

**Comment by reflog** [comment]
@mkraft - the problem here is that Jenkins scripts doesn't give access to 'docker' executable. I guess I need to run sub task with DnD (docker-in-docker). That's why I invoked @cpanato  :) Locally everything works fine.

**Comment by cpanato** [comment]
@reflog @mkraft the tests are running inside the docker, so you cannot invoke docker commands in that step or you somehow run outside or change to use the `mysql` or `postgres` dns names that are available since we create a network for them

**Comment by cpanato** [comment]
weird is why this did not set the build to fail :/

**Comment by cpanato** [comment]
invoked sounds very magical
/gandalf

**Comment by reflog** [comment]
@cpanato - thanks for the explaination. How can we add `make test-db-migration` as a required CI step?

**Comment by cpanato** [comment]
@reflog you can change the jenkinsfile.pr to achieve that

**Comment by reflog** [comment]
@cpanato - I'd appreciate some guidance on how to do that

**Comment by cpanato** [comment]
lets discuss in the community, ping me

File: store/sqlstore/upgrade.go [hunk 1]
```diff
@@ -688,10 +688,10 @@ func UpgradeDatabaseToVersion511(sqlStore SqlStore) {
 
 func UpgradeDatabaseToVersion512(sqlStore SqlStore) {
 	if shouldPerformUpgrade(sqlStore, VERSION_5_11_0, VERSION_5_12_0) {
-		sqlStore.CreateColumnIfNotExistsNoDefault("TeamMembers", "SchemeGuest", "boolean", "boolean")
-		sqlStore.CreateColumnIfNotExistsNoDefault("ChannelMembers", "SchemeGuest", "boolean", "boolean")
-		sqlStore.CreateColumnIfNotExistsNoDefault("Schemes", "DefaultTeamGuestRole", "text", "VARCHAR(64)")
-		sqlStore.CreateColumnIfNotExistsNoDefault("Schemes", "DefaultChannelGuestRole", "text", "VARCHAR(64)")
+		sqlStore.CreateColumnIfNotExistsNoDefault("TeamMembers", "SchemeGuest", "tinyint(4)", "boolean")
```
Comment (lieut-data) [inline store/sqlstore/upgrade.go]: Retroactively updating this migration won't help anyone who's already upgraded to 5.12 or 5.13: should we move this to `UpgradeDatabaseToVersion516`? Or `UpgradeDatabaseToVersion514` and backport to the active release?

Comment (reflog) [reply store/sqlstore/upgrade.go]: @lieut-data - I've tried adding it to `upgrade.go` but it still resuls it:
```
Generating diff
210,211c210
<   PRIMARY KEY (`GroupId`,`ChannelId`),
<   KEY `idx_groupchannels_channelid` (`ChannelId`)
---
>   PRIMARY KEY (`GroupId`,`ChannelId`)
234,235c233
<   PRIMARY KEY (`GroupId`,`TeamId`),
<   KEY `idx_groupteams_teamid` (`TeamId`)
---
>   PRIMARY KEY (`GroupId`,`TeamId`)
```

Comment (lieut-data) [reply store/sqlstore/upgrade.go]: I admit I'm not sure how that could happen. cc @mkraft for clarity, but I maintain we need to do this in `UpgradeDatabaseToVersion514` for the reasons stated above.

Comment (reflog) [reply store/sqlstore/upgrade.go]: @lieut-data - you are correct of course. I've fixed up the order of the applied operations. It's fine now.

Comment (lieut-data) [reply store/sqlstore/upgrade.go]: @reflog, apologies for any confusion here -- but I'm still not seeing these indices being re-created in the `UpgradeDatabaseToVersion514`. The DB migration will pass now, but any customer that happened to do a fresh install of 5.10-5.13 won't have gotten these indexes, and won't ever get them unless we add them to a migration step.

Comment (reflog) [reply store/sqlstore/upgrade.go]: @lieut-data - the indices are added here in `upgrade.go`:
```go
func UpgradeDatabaseToVersion510(sqlStore SqlStore) {
	if shouldPerformUpgrade(sqlStore, VERSION_5_9_0, VERSION_5_10_0) {
		sqlStore.CreateColumnIfNotExistsNoDefault("Channels", "GroupConstrained", "tinyint(4)", "boolean")
		sqlStore.CreateColumnIfNotExistsNoDefault("Teams", "GroupConstrained", "tinyint(4)", "boolean")

		sqlStore.CreateIndexIfNotExists("idx_groupteams_teamid", "GroupTeams", "TeamId")
		sqlStore.CreateIndexIfNotExists("idx_groupchannels_channelid", "GroupChannels", "ChannelId")

		saveSchemaVersion(sqlStore, VERSION_5_10_0)
	}
}
```

Comment (reflog) [reply store/sqlstore/upgrade.go]: ok, spoke to @lieut-data , realized what I was missing. Added the indices to 514 migration to make sure they are performed.

**Comment by cpanato** [comment]
merged before my review, removing myself

**Comment by cpanato** [comment]
@lieut-data there are new things in the pipeline, please take a look as well

**Pull request merged**